src/Xrefcheck/System.hs

```diff
@@ -174,27 +171,53 @@ mkGlobPattern path = do
         https://hackage.haskell.org/package/Glob/docs/System-FilePath-Glob.html#v:compile
         Special characters in file names can be escaped using square brackets, e.g. <a> -> [<]a[>].
         |]
-
-bindGlobPattern :: CanonicalPath -> RelGlobPattern -> Glob.Pattern
-bindGlobPattern root (UnsafeRelGlobPattern relPat) = readingSystem $ do
-  UnsafeCanonicalPath absPat <- root </ relPat
-  case Glob.tryCompileWith globCompileOptions absPat of
-    Left err -> error $
-      "Glob pattern compilation failed after canonicalization: " <> toText err
-    Right pat ->
-      return pat
-
-matchesGlobPatterns :: CanonicalPath -> [RelGlobPattern] -> CanonicalPath -> Bool
-matchesGlobPatterns root globPatterns file = or
-  [ Glob.match pat $ unCanonicalPath file
-  | globPattern <- globPatterns
-  , let pat = bindGlobPattern root globPattern
+  where
+    normalise = toString
+      . expandPosixIndirections
+      . fromString
+      . FPP.normalise
+      . FPP.dropTrailingPathSeparator
+
+-- Checks if a 'CanonicalRelPosixLink' matches some of the given
+-- 'CanonicalRelGlobPattern's.
+--
+-- They are considered as relative to the same root.
+matchesGlobPatterns :: [CanonicalRelGlobPattern] -> CanonicalRelPosixLink -> Bool
```

Comment:
Glob patterns are canonicalised and matched against canonicalised links.

I tried to not canonicalise glob patterns or to not canonicalise the links that we match for, but we have test cases that require either one or the other thing.

Reply:
Could you elaborate what problems we get if glob is not normalized?

Reply:
Yeah! I have found the test cases that fail.

### If we do not normalize Glob patterns

In `check-local-refs/Checking behavior when there are virtual files, root = "dir1"`, the following links are supposed to be ignored by config:

A) `dir2/../DIR2` (canonicalized is DIR2)
B) `b/../../b.md` (canonicalized is ../b.md)

Config indicates if with the following patterns:

`DIR2/` (this will match A if we remove trailing whitespaces in the pattern)
`b/../../*` (this will match B if we expand indirections in the pattern)

### If we do not normalize neither Glob patterns nor the paths that we try to match

Given the previous case, we can think about trying to match paths as they are, but then other test case fails

In `check-ignoreLocalRefsTo/IgnoreLocalRefsTo: all references should be valid`, the following link is supposed to be ignored by config:

C) `one/../two/b.md`

Config indicates it with the following pattern:

`./two/*` (this will match C if indirections are expanded both in the pattern and in C itself)

So, our test cases suggest that we expect glob matching to be normalized in both sides: patterns and links.

Reply:
Thanks, this clarifies to me, sounds like a fair change :+1: